### PR TITLE
DB-10819 Fix inconsistent results with substr between spark and native spark 

### PR DESCRIPTION
**Pull request by arnaud-splice**

### CODE REVIEW

**Comment by arnaud-splice** [comment]
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

**Review by msirek — APPROVED** [review]
Looks good.  Just check the parameters to substring and add the comma if required.

**Comment by spliceci** [comment]
TEST SUCCEEDED +1

File: db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/OperatorToString.java [hunk 1]
```diff
@@ -679,23 +679,26 @@ else if (vars.buildExpressionTree)
...
                         String retval = format("%s(%s, %s, %s) ", top.getOperator(), opToString2(top.getReceiver(), vars),
                                 opToString2(top.getLeftOperand(), vars), opToString2(top.getRightOperand(), vars));
                         vars.relationalOpDepth.decrement();
                         return retval;
                     } else if (top.getOperator().equals("substring")) {
-                        assert top.getRightOperand() == null;
                         vars.relationalOpDepth.decrement();
-                        String retval = format("%s(%s, %s) ", top.getOperator(), opToString2(top.getReceiver(), vars),
-                                opToString2(top.getLeftOperand(), vars));
+                        String retval = format("%s(%s, %s %s)",
```
[the remark this answers is not in the document]
Comment (arnaud-splice) [reply db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/OperatorToString.java:690]: No, the comma is added if a third parameter is necessary, see below.

[the remark this answers is not in the document]
Comment (msirek) [reply db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/OperatorToString.java:690]: OK, I see.  Thanks.

File: splice_machine/src/test/java/com/splicemachine/derby/test/framework/SpliceUnitTest.java [hunk 2]
```diff
@@ -1195,6 +1195,21 @@ protected void checkBooleanExpression(String input, boolean output, TestConnecti
...
             Assert.assertEquals(output, rs.getBoolean(1));
         }
+    }
+
+    protected void checkExpressionType(String input, String expectedType, TestConnection conn) throws SQLException {
+        String sql = format("select typeof(%s)", input);
+        try(ResultSet rs = conn.query(sql)) {
+            rs.next();
+            Assert.assertEquals(expectedType, rs.getString(1));
+        }
```
Comment (ascend1) [inline splice_machine/src/test/java/com/splicemachine/derby/test/framework/SpliceUnitTest.java:1205]: This logic is available as `firstRowContainsQuery()` in the same file. But it's just a minor issue and should not prevent the change from being merged.

File: splice_machine/src/test/java/com/splicemachine/derby/test/framework/SpliceUnitTest.java [hunk 3]
```diff
@@ -1195,6 +1195,21 @@ protected void checkBooleanExpression(String input, boolean output, TestConnecti
...
+    }
 
+    protected void checkStringExpression(String input, String expectedOutput, TestConnection conn) throws SQLException {
+        String sql = format("select %s", input);
+        try(ResultSet rs = conn.query(sql)) {
+            rs.next();
+            Assert.assertEquals(expectedOutput, rs.getString(1));
+        }
```
Comment (ascend1) [inline splice_machine/src/test/java/com/splicemachine/derby/test/framework/SpliceUnitTest.java:1213]: Ditto.

**Review by ascend1 — APPROVED** [review]

**Review by dgomezferro — APPROVED** [review]

**Review by carolp-503 — APPROVED** [review]
+1

**Pull request merged**